packages/graphql/src/api-v6/queryIRFactory/FilterFactory.ts

```diff
@@ -56,176 +57,85 @@ export class FilterFactory {
-    }
-
-    public createTopLevelFilters({
-        where = {},
-        entity,
-    }: {
-        entity: ConcreteEntity;
-        where?: GraphQLWhereTopLevel;
-    }): Filter[] {
-        const andFilters = this.createTopLevelLogicalFilters({ operation: "AND", entity, where: where.AND });
-        const orFilters = this.createTopLevelLogicalFilters({ operation: "OR", entity, where: where.OR });
-        const notFilters = this.createTopLevelLogicalFilters({
-            operation: "NOT",
-            entity,
-            where: where.NOT ? [where.NOT] : undefined,
-        });
-
-        const edgeFilters = this.createNodeFilter({ entity, where: where.node });
-        return this.mergeFiltersWithAnd([
-            ...this.mergeFiltersWithAnd(edgeFilters),
-            ...andFilters,
-            ...orFilters,
-            ...notFilters,
-        ]);
-    }
-
-    private createTopLevelLogicalFilters({
-        where = [],
-        operation,
-        entity,
-    }: {
-        entity: ConcreteEntity;
-        operation: LogicalOperators;
-        where?: GraphQLWhereTopLevel[];
-    }): [] | [Filter] {
-        if (where.length === 0) {
-            return [];
-        }
-        const nestedFilters = where.flatMap((orWhere: GraphQLWhereTopLevel) => {
-            return this.createTopLevelFilters({ entity, where: orWhere });
-        });
+        where?: GraphQLWhere | GraphQLWhereTopLevel | GraphQLEdgeWhere;
+    }): Filter | undefined {
+        const filters = filterTruthy(
+            Object.entries(where).map(([key, value]) => {
+                if (key === "AND" || key === "OR" || key === "NOT") {
+                    const whereInLogicalOperation = asArray(value) as Array<
+                        GraphQLWhere | GraphQLWhereTopLevel | GraphQLEdgeWhere
+                    >;
+                    const nestedFilters = whereInLogicalOperation.map((nestedWhere) => {
+                        return this.createFilters({
+                            where: nestedWhere,
+                            relationship,
+                            entity,
+                        });
+                    });
+
+                    return new LogicalFilter({
+                        operation: key,
+                        filters: filterTruthy(nestedFilters),
+                    });
+                }
+
+                if (key === "edges") {
+                    return this.createFilters({ entity, relationship, where: value });
+                }
+
+                if (key === "node") {
+                    return this.createNodeFilter({ entity, where: value });
+                }
+
+                if (key === "properties" && relationship) {
+                    const edgePropertiesFilters = this.createEdgePropertiesFilters({
+                        where: value,
+                        relationship,
+                    });
+
+                    return new LogicalFilter({
```

Comment:
IMO this logical operator logic should be defined inside `createEdgePropertiesFilters` as is done for `createNodeFilters`. 
I would personally avoid having methods that return `Filter[]` as I would wrap these filters in a `LogicalFilter`.
For instance, here you would have the `LogicalFilter` broken for relationship properties.
Try this:
```gql
query {
    movies {
        connection {
            edges {
                node {
                    title
                    actors(
                        where: {
                            edges: {
                                properties: { NOT: { year: { equals: 1999 } } }
                            }
                        }
                    ) {
                        connection {
                            edges {
                                node {
                                    name
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}
```